stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4660,48 +4660,80 @@ LogicalResult ReplicaIdOp::inferReturnTypes(
   return success();
 }
 
-LogicalResult IfOp::verify() {
-  if (failed(verifyConditionalBranch(*this, true_branch(),
-                                     /*branchName=*/"true_branch"))) {
-    return failure();
+bool isCompatibleForTypeRange(ValueTypeRange<OperandRange> typeRange1,
+                              ValueTypeRange<OperandRange> typeRange2) {
+  if (typeRange1.size() != typeRange2.size()) return false;
+  for (auto it : llvm::zip(typeRange1, typeRange2)) {
+    if (!mlir::hlo::isCompatibleForHloTypeInference(std::get<0>(it),
+                                                    std::get<1>(it)))
+      return false;
   }
+  return true;
+}
 
-  if (failed(verifyConditionalBranch(*this, false_branch(),
-                                     /*branchName=*/"false_branch"))) {
+LogicalResult IfOp::inferReturnTypeComponents(
+    MLIRContext*, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  IfOp::Adaptor adaptor(operands, attributes, regions);
+  if (failed(verifyConditionalBranch(adaptor.true_branch(),
+                                     /*branchName=*/"true_branch", location)))
     return failure();
-  }
+  if (failed(verifyConditionalBranch(adaptor.false_branch(),
+                                     /*branchName=*/"false_branch", location)))
+    return failure();
+
+  auto trueBranchRetureTypes =
+      adaptor.true_branch().front().getTerminator()->getOperandTypes();
+  auto falseBranchRetureTypes =
+      adaptor.false_branch().front().getTerminator()->getOperandTypes();
+  
+  if (!isCompatibleForTypeRange(trueBranchRetureTypes, falseBranchRetureTypes))
+    return emitOptionalError(
+        location, "true_branch and false_branch mismatched return types: ",
+        trueBranchRetureTypes, " vs ", falseBranchRetureTypes);
+
+  for (auto trueBranchRetureType : trueBranchRetureTypes)
+    inferredReturnShapes.emplace_back(trueBranchRetureType.cast<ShapedType>());
   return success();
 }
 
 //===----------------------------------------------------------------------===//
 // Case Op
 //===----------------------------------------------------------------------===//
 
-LogicalResult CaseOp::verify() {
-  auto numBranches = branches().size();
-
-  for (unsigned i = 0; i < numBranches; ++i)
-    if (failed(verifyConditionalBranch(*this, branches()[i],
-                                       /*branchName=*/"branch " + Twine(i))))
+LogicalResult CaseOp::inferReturnTypeComponents(
+    MLIRContext*, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  CaseOp::Adaptor adaptor(operands, attributes, regions);
+  ValueTypeRange<OperandRange> branch0ResultTypes =
+      adaptor.branches()[0]->front().getTerminator()->getOperandTypes();
```

Comment:
We should check that the number of branches is non-zero.

Reply:
Good catch, added in the tests.